openwrt_api.py

```diff
@@ -0,0 +1,400 @@
+"""
+   Author: Xuefeng Huang
+
+   Start Date: 15 August 2015
+
+   Description:
+
+   This file provides a python interface to low-level system call or files 
+   on the OpenWRT platform. It is designed to let researchers do a wide range 
+   of network measurements on home gateway.
+   
+"""
+
+import nanny
+
+import socket
+
+import portable_popen # Import for Popen
+
+import textops # Import seattlelib's text processing lib
+
+import re
+
+import os # Provides some convenience functions
+
+import emulcomm
+
+# Get the exceptions
+from exception_hierarchy import *
+
+safe_open = open
+
+##### Public Functions #####     
+def get_network_bytes(interface):
+  """
+  <Purpose>
+    Count the number of packets received and transmitted from an Interface.
+
+  <Arguments>
+    interface: the name of the interface to gather network information.
+
+  <Exceptions>
+    RepyArgumentError is raised if the interface aren't valid types or values
+    FileNotFoundError is raised if the file does not exist
+
+  <Side Effects>
+    None
+
+  <Returns>
+    Network usage info as a dict such as {'trans': '59528148', 'recv': '1366399094'}
+  """
+
+  if type(interface) is not str:
+    raise RepyArgumentError("get_network_bytes() takes a string as argument.")
+
+  # Check the input arguments (sanity)
+  if interface not in get_network_interface():
+    raise RepyArgumentError("interface "+ interface + " is not available.")
+
+  if os.path.exists("/proc/net/dev"):
+    dev = safe_open("/proc/net/dev", "r")
+    for line in dev:
+      if interface in line:
+        data = line.split('%s:' % interface)[1].split()
+        rx_bytes, tx_bytes = (data[0], data[8])
+        return {"recv": rx_bytes, "trans": tx_bytes}
+  else:
+    raise FileNotFoundError("Could not find /proc/net/dev!")
+
+def get_network_packets(interface):
+  """
+  <Purpose>
+    Count the number of packets received and transmitted from an interface
+
+  <Arguments>
+    interface: the name of the interface to gather network information.
+
+  <Exceptions>
+    RepyArgumentError is raised if the interface aren't valid types or values
+    FileNotFoundError is raised if the file does not exist
+
+  <Side Effects>
+    None
+
+  <Returns>
+    Network usage info as a dict such as {'trans': '598602', 'recv': '1262217'}
+  """
+  if type(interface) is not str:
+    raise RepyArgumentError("get_network_bytes() takes a string as argument.")
+  
+  # Check the input arguments (sanity)
+  if interface not in get_network_interface():
+    raise RepyArgumentError("interface "+ interface + " is not available.")
+
+  if os.path.exists("/proc/net/dev"):
+    dev = safe_open('/proc/net/dev', 'r')
+    for line in dev:
+      if interface in line:
+        data = line.split('%s:' % interface)[1].split()
+        rx_packets, tx_packets = (data[1], data[9])
```

Comment:
Almost but not quite the same splitting / looping as above. How about writing one (internal) function, e.g. `_get_interface_traffic_statistics`, that calls into `netstat`and does the parsing, then have two separate high-level functions that return byte/packet-related partial information?